File: prayer_app_expo/src/screens/QuranSurahListScreen.js
import React, { useEffect, useMemo, useState } from 'react';
import {
    View,
    Text,
    StyleSheet,
    TouchableOpacity,
    TextInput,
    FlatList,
    ActivityIndicator,
} from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { useTheme } from '../providers/ThemeProvider';
import { QuranLayout, interFont, getTypography } from '../theme/theme';
import { fetchSurahList, loadCachedSurahList } from '../services/quranApi';
import { loadLastRead } from '../services/quranStorageService';

export default function QuranSurahListScreen({ onBack, onOpenReader, autofocusSearch = false }) {
    const { theme: tc } = useTheme();
    const typo = getTypography(tc);

    const [query, setQuery] = useState('');
    const [surahs, setSurahs] = useState([]);
    const [loading, setLoading] = useState(true);
    const [offlineCached, setOfflineCached] = useState(false);
    const [error, setError] = useState(null);
    const [lastRead, setLastRead] = useState(null);

    useEffect(() => {
        loadData();
        refreshLastRead();
    }, []);

    async function refreshLastRead() {
        const p = await loadLastRead();
        setLastRead(p);
    }

    async function loadData() {
        setLoading(true);
        setOfflineCached(false);
        setError(null);

        const cached = await loadCachedSurahList();
        if (cached?.length) {
            setSurahs(cached);
            setLoading(false);
        }

        try {
            const fresh = await fetchSurahList();
            setSurahs(fresh);
            setLoading(false);
            setOfflineCached(false);
        } catch (_) {
            if (cached?.length) {
                setOfflineCached(true);
                setLoading(false);
            } else {
                setError('Failed to load surah list');
                setLoading(false);
            }
        }
    }

    const filtered = useMemo(() => {
        const qRaw = query.trim();
        const q = qRaw.toLowerCase();
        if (!qRaw) return surahs;
        return surahs.filter((s) =>
            String(s.number).includes(q) ||
            (s.englishName || '').toLowerCase().includes(q) ||
            (s.nameAr || '').includes(qRaw)
        );
    }, [query, surahs]);

    return (
        <View style={{ flex: 1 }}>
            <View style={styles.topBar}>
                <TouchableOpacity onPress={onBack} hitSlop={10}>
                    <MaterialCommunityIcons name="arrow-left" size={24} color={tc.textPrimary} />
                </TouchableOpacity>
                <View style={{ width: 12 }} />
                <Text style={typo.titleMedium}>All Surahs</Text>
            </View>

            <View style={{ paddingHorizontal: QuranLayout.screenPadding }}>
                <View style={[styles.searchWrap, { backgroundColor: tc.card, borderColor: tc.cardBorder }]}> 
                    <MaterialCommunityIcons name="magnify" size={20} color={tc.textMuted} />
                    <View style={{ width: 8 }} />
                    <TextInput
                        value={query}
                        onChangeText={setQuery}
                        autoFocus={autofocusSearch}
                        placeholder="Search by number, English, or Arabic"
                        placeholderTextColor={tc.textMuted}
                        style={[styles.searchInput, { color: tc.textPrimary }]}
                    />
                </View>
            </View>

            {offlineCached && (
                <View style={[styles.banner, { backgroundColor: tc.card, borderColor: tc.cardBorder }]}> 
                    <MaterialCommunityIcons name="wifi-off" size={16} color={tc.textMuted} />
                    <Text style={[typo.caption, { marginLeft: 8 }]}>Offline (cached)</Text>
                </View>
            )}

            {loading && surahs.length === 0 ? (
                <View style={styles.center}><ActivityIndicator size="large" color={tc.accent} /></View>
            ) : error && surahs.length === 0 ? (
                <View style={styles.center}><Text style={typo.caption}>{error}</Text></View>
            ) : (
                <FlatList
                    data={filtered}
                    keyExtractor={(item) => String(item.number)}
                    contentContainerStyle={{ paddingHorizontal: 20, paddingTop: 10, paddingBottom: 20 }}
                    renderItem={({ item }) => {
                        const isLastRead = lastRead?.surahNumber === item.number;
                        const initialAyah = isLastRead ? lastRead.ayahNumber : 1;
                        return (
                            <TouchableOpacity
                                activeOpacity={0.8}
                                onPress={async () => {
                                    await onOpenReader(item, initialAyah, 'list');
                                    refreshLastRead();
                                }}
                                style={[
                                    styles.row,
                                    {
                                        backgroundColor: tc.card,
                                        borderColor: isLastRead ? `${tc.accent}AA` : tc.cardBorder,
                                    },
                                ]}
                            >
                                <View style={[styles.badge, { backgroundColor: `${tc.accent}2E` }]}> 
                                    <Text style={[typo.caption, { color: tc.accent, fontFamily: interFont('600') }]}>{item.number}</Text>
                                </View>
                                <View style={{ width: 12 }} />
                                <View style={{ flex: 1 }}>
                                    <Text style={[typo.body, { fontFamily: interFont('600') }]} numberOfLines={1}>{item.englishName}</Text>
                                    <Text style={typo.caption} numberOfLines={1}>{item.nameAr}</Text>
                                </View>
                                <View style={{ marginLeft: 8, alignItems: 'flex-end' }}>
                                    <Text style={typo.caption}>{item.ayahCount} ayahs</Text>
                                    <Text style={typo.caption}>{item.revelationType}</Text>
                                </View>
                                <View style={{ width: 4 }} />
                                <MaterialCommunityIcons name="chevron-right" size={20} color={tc.textMuted} />
                            </TouchableOpacity>
                        );
                    }}
                />
            )}
        </View>
    );
}

const styles = StyleSheet.create({
    topBar: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingHorizontal: 12,
        paddingVertical: 8,
    },
    searchWrap: {
        height: QuranLayout.searchHeight,
        borderWidth: 1,
        borderRadius: QuranLayout.searchRadius,
        flexDirection: 'row',
        alignItems: 'center',
        paddingHorizontal: 12,
    },
    searchInput: {
        flex: 1,
        fontFamily: interFont('400'),
        fontSize: 14,
    },
    banner: {
        marginHorizontal: 20,
        marginTop: 8,
        borderWidth: 1,
        borderRadius: 10,
        paddingHorizontal: 12,
        paddingVertical: 8,
        flexDirection: 'row',
        alignItems: 'center',
    },
    center: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },
    row: {
        height: QuranLayout.surahRowHeight,
        marginBottom: 10,
        borderRadius: QuranLayout.cardRadius,
        borderWidth: 1,
        paddingHorizontal: 12,
        flexDirection: 'row',
        alignItems: 'center',
    },
    badge: {
        width: 32,
        height: 32,
        borderRadius: 16,
        alignItems: 'center',
        justifyContent: 'center',
    },
});

File: prayer_app_flutter/lib/src/screens/quran_surah_list_screen.dart
import 'package:flutter/material.dart';
import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';
import '../models/quran_models.dart';
import '../services/quran_api_service.dart';
import '../services/quran_storage_service.dart';
import '../theme/app_theme.dart';
import '../providers/theme_provider.dart';
import 'quran_reader_screen.dart';

class QuranSurahListScreen extends StatefulWidget {
  final bool autofocusSearch;

  const QuranSurahListScreen({super.key, this.autofocusSearch = false});

  @override
  State<QuranSurahListScreen> createState() => _QuranSurahListScreenState();
}

class _QuranSurahListScreenState extends State<QuranSurahListScreen> {
  final QuranApiService _api = QuranApiService();
  final QuranStorageService _storage = QuranStorageService();
  final TextEditingController _searchCtrl = TextEditingController();

  List<SurahMeta> _surahs = [];
  QuranPointer? _lastRead;
  bool _loading = true;
  bool _offlineCached = false;
  String? _error;

  @override
  void initState() {
    super.initState();
    _loadLastRead();
    _loadSurahs();
    _searchCtrl.addListener(() => setState(() {}));
  }

  @override
  void dispose() {
    _searchCtrl.dispose();
    super.dispose();
  }

  Future<void> _loadLastRead() async {
    final p = await _storage.loadLastRead();
    if (!mounted) return;
    setState(() => _lastRead = p);
  }

  Future<void> _loadSurahs() async {
    setState(() {
      _loading = true;
      _error = null;
      _offlineCached = false;
    });

    final cached = await _api.loadCachedSurahList();
    if (cached != null && cached.isNotEmpty && mounted) {
      setState(() {
        _surahs = cached;
        _loading = false;
      });
    }

    try {
      final fresh = await _api.fetchSurahList();
      if (!mounted) return;
      setState(() {
        _surahs = fresh;
        _loading = false;
        _offlineCached = false;
      });
    } catch (e) {
      if (!mounted) return;
      if (cached != null && cached.isNotEmpty) {
        setState(() {
          _offlineCached = true;
          _loading = false;
        });
      } else {
        setState(() {
          _error = 'Could not load surah list. Check internet and retry.';
          _loading = false;
        });
      }
      debugPrint('[QuranList] fetch failed: $e');
    }
  }

  List<SurahMeta> get _filtered {
    final qRaw = _searchCtrl.text.trim();
    final q = qRaw.toLowerCase();
    if (qRaw.isEmpty) return _surahs;
    return _surahs.where((s) {
      return s.number.toString().contains(q) ||
          s.englishName.toLowerCase().contains(q) ||
          s.nameAr.contains(qRaw);
    }).toList();
  }

  Future<void> _openReader(SurahMeta surah) async {
    final initialAyah = _lastRead?.surahNumber == surah.number
        ? _lastRead!.ayahNumber
        : 1;
    await Navigator.of(context).push(
      MaterialPageRoute(
        builder: (_) =>
            QuranReaderScreen(surah: surah, initialAyahNumber: initialAyah),
      ),
    );
    _loadLastRead();
  }

  @override
  Widget build(BuildContext context) {
    final tc = ThemeScope.of(context).current;

    return Scaffold(
      body: Container(
        decoration: BoxDecoration(gradient: appBackgroundGradient(tc)),
        child: SafeArea(
          bottom: false,
          child: Column(
            children: [
              Padding(
                padding: const EdgeInsets.fromLTRB(8, 8, 16, 8),
                child: Row(
                  children: [
                    IconButton(
                      onPressed: () => Navigator.of(context).pop(),
                      icon: Icon(MdiIcons.arrowLeft, color: tc.textPrimary),
                    ),
                    Expanded(
                      child: Text(
                        'All Surahs',
                        style: AppTypography.titleMedium(tc),
                      ),
                    ),
                  ],
                ),
              ),
              Padding(
                padding: const EdgeInsets.symmetric(
                  horizontal: QuranLayout.screenPadding,
                ),
                child: Container(
                  height: QuranLayout.searchHeight,
                  decoration: BoxDecoration(
                    color: tc.card,
                    borderRadius: BorderRadius.circular(
                      QuranLayout.searchRadius,
                    ),
                    border: Border.all(color: tc.cardBorder),
                  ),
                  child: Row(
                    children: [
                      const SizedBox(width: 12),
                      Icon(MdiIcons.magnify, color: tc.textMuted, size: 20),
                      const SizedBox(width: 8),
                      Expanded(
                        child: TextField(
                          controller: _searchCtrl,
                          autofocus: widget.autofocusSearch,
                          style: AppTypography.body(tc).copyWith(fontSize: 14),
                          decoration: InputDecoration(
                            border: InputBorder.none,
                            hintText: 'Search by number, English, or Arabic',
                            hintStyle: AppTypography.caption(tc),
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
              if (_offlineCached)
                Container(
                  margin: const EdgeInsets.fromLTRB(20, 8, 20, 0),
                  padding: const EdgeInsets.symmetric(
                    horizontal: 12,
                    vertical: 8,
                  ),
                  decoration: BoxDecoration(
                    color: tc.card,
                    borderRadius: BorderRadius.circular(10),
                    border: Border.all(color: tc.cardBorder),
                  ),
                  child: Row(
                    children: [
                      Icon(MdiIcons.wifiOff, size: 16, color: tc.textMuted),
                      const SizedBox(width: 8),
                      Text(
                        'Offline (cached)',
                        style: AppTypography.caption(tc),
                      ),
                    ],
                  ),
                ),
              Expanded(
                child: _loading && _surahs.isEmpty
                    ? Center(child: CircularProgressIndicator(color: tc.accent))
                    : _error != null && _surahs.isEmpty
                    ? Center(
                        child: Padding(
                          padding: const EdgeInsets.symmetric(horizontal: 24),
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              Text(
                                _error!,
                                textAlign: TextAlign.center,
                                style: AppTypography.caption(tc),
                              ),
                              const SizedBox(height: 12),
                              OutlinedButton(
                                onPressed: _loadSurahs,
                                child: const Text('Retry'),
                              ),
                            ],
                          ),
                        ),
                      )
                    : ListView.builder(
                        itemCount: _filtered.length,
                        padding: const EdgeInsets.fromLTRB(20, 10, 20, 20),
                        itemBuilder: (context, index) {
                          final s = _filtered[index];
                          final isLastRead = _lastRead?.surahNumber == s.number;
                          return GestureDetector(
                            onTap: () => _openReader(s),
                            child: Container(
                              height: QuranLayout.rowHeight,
                              margin: const EdgeInsets.only(bottom: 10),
                              padding: const EdgeInsets.symmetric(
                                horizontal: 12,
                              ),
                              decoration: BoxDecoration(
                                color: tc.card,
                                borderRadius: BorderRadius.circular(
                                  QuranLayout.cardRadius,
                                ),
                                border: Border.all(
                                  color: isLastRead
                                      ? tc.accent.withValues(alpha: 0.65)
                                      : tc.cardBorder,
                                ),
                              ),
                              child: Row(
                                children: [
                                  Container(
                                    width: 32,
                                    height: 32,
                                    alignment: Alignment.center,
                                    decoration: BoxDecoration(
                                      color: tc.accent.withValues(alpha: 0.18),
                                      borderRadius: BorderRadius.circular(16),
                                    ),
                                    child: Text(
                                      '${s.number}',
                                      style: AppTypography.caption(tc).copyWith(
                                        color: tc.accent,
                                        fontWeight: FontWeight.w600,
                                      ),
                                    ),
                                  ),
                                  const SizedBox(width: 12),
                                  Expanded(
                                    child: Column(
                                      mainAxisAlignment:
                                          MainAxisAlignment.center,
                                      crossAxisAlignment:
                                          CrossAxisAlignment.start,
                                      children: [
                                        Text(
                                          s.englishName,
                                          style: AppTypography.body(tc)
                                              .copyWith(
                                                fontWeight: FontWeight.w600,
                                              ),
                                          maxLines: 1,
                                          overflow: TextOverflow.ellipsis,
                                        ),
                                        Text(
                                          s.nameAr,
                                          style: AppTypography.caption(tc),
                                          textDirection: TextDirection.rtl,
                                          maxLines: 1,
                                          overflow: TextOverflow.ellipsis,
                                        ),
                                      ],
                                    ),
                                  ),
                                  const SizedBox(width: 8),
                                  Column(
                                    mainAxisAlignment: MainAxisAlignment.center,
                                    crossAxisAlignment: CrossAxisAlignment.end,
                                    children: [
                                      Text(
                                        '${s.ayahCount} ayahs',
                                        style: AppTypography.caption(tc),
                                      ),
                                      Text(
                                        s.revelationType,
                                        style: AppTypography.caption(tc),
                                      ),
                                    ],
                                  ),
                                  const SizedBox(width: 4),
                                  Icon(
                                    MdiIcons.chevronRight,
                                    color: tc.textMuted,
                                  ),
                                ],
                              ),
                            ),
                          );
                        },
                      ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
